gulpfile.js

```diff
@@ -0,0 +1,381 @@
+const karma = require('karma');
+
+// Gulp plugins.
+const gulpClean = require('gulp-clean');
+const gulpServer = require('gulp-server-livereload');
+const gulpMerge = require('merge2');
+const gulpRunSequence = require('run-sequence');
+const gulpSass = require('gulp-sass');
+const gulpSourcemaps = require('gulp-sourcemaps');
+const gulpTs = require('gulp-typescript');
+
+
+// Directories.
+const srcDir = path.join(__dirname, 'src');
+const componentsDir = path.join(srcDir, 'lib');
+const devAppDir = path.join(srcDir, 'demo-app');
+const e2eAppDir = path.join(srcDir, 'e2e-app');
+
+const outDir = 'dist';
+const outLibDir = path.join(outDir, '@angular2-material');
+
+
+/**
+ * Create a TS Build Task, based on the options.
+ */
+function createTsBuildTask(options) {
+  const tsConfigDir = options.tsConfigPath;
+  const tsConfigPath = path.join(tsConfigDir, 'tsconfig.json');
+  const tsConfig = JSON.parse(fs.readFileSync(tsConfigPath, 'utf-8'));
+  const dest = path.join(tsConfigDir, tsConfig.compilerOptions.outDir);
+
+  return function() {
+    const tsProject = gulpTs.createProject(tsConfigPath, {
+      typescript: require('typescript')
+    });
+
+    let pipe = tsProject.src()
+      .pipe(gulpSourcemaps.init())
+      .pipe(gulpTs(tsProject));
+    let dts = pipe.dts.pipe(gulp.dest(dest));
+
+    return gulpMerge([
+      dts,
+      pipe
+        .pipe(gulpSourcemaps.write('.'))
+        .pipe(gulp.dest(dest))
+    ]);
+  };
+}
+
+/**
+ * Create a SASS Build Task.
+ */
+function createSassBuildTask(options) {
+  const dest = options.dest;
+  const glob = path.join(options.root, '**/*.scss');
+  const sassOptions = {
+    includePaths: options.includePaths
+  };
+
+  return function() {
+    return gulp.src(glob)
+      .pipe(gulpSourcemaps.init())
+      .pipe(gulpSass(sassOptions).on('error', gulpSass.logError))
+      .pipe(gulpSourcemaps.write(dest))
+      .pipe(gulp.dest(dest));
+  };
+}
+
+
+/**
+ * Create a Gulp task that executes a process.
+ */
+function createExecTask(packageName, executable, args) {
+  if (!args) {
+    args = executable;
+    executable = undefined;
+  }
+  return function(done) {
+    resolveBin(packageName, { executable: executable }, function (err, binPath) {
+      if (err) {
+        console.error(err);
+        throw err;
+      }
+
+      const spawn = child_process.spawn(binPath, args);
+
+      spawn.stdout.on('data', (data) => {
+        process.stdout.write(data);
+      });
+
+      spawn.stderr.on('data', (data) => {
+        process.stderr.write(data);
+      });
+
+      spawn.on('close', (code) => {
+        if (code != 0) {
+          throw new Error('Process failed with code ' + code);
+        }
+        done();
+      });
+    });
+  }
+}
+
+/***************************************************************************************************
+ * Components Build Tasks.
+ */
+gulp.task(':build:components:ts', createTsBuildTask({ tsConfigPath: componentsDir }));
```

Comment:
I don't quite follow your task naming convention. I'm guessing the leading colon means it's "private" (only meant to be consumed as the dep of another rule)? 

What do you think of using `_` instead of a leading colon and also using dots instead of colons for semantic delimitiers, e.g. `_build.components.assets`. I find the dots to be easier to read and to type.